src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,728 @@
+		return -EINVAL;
+	}
+
+	if (!format_is_supported(src_fmt->frame_fmt)) {
+		comp_err(dev, "waves_effect_check() float samples not supported");
+		return -EINVAL;
+	}
+
+	if (!layout_is_supported(component->ca_source->buffer_fmt)) {
+		comp_err(dev, "waves_effect_check() non interleaved not supported");
+		return -EINVAL;
+	}
+
+	if (!rate_is_supported(src_fmt->rate)) {
+		comp_err(dev, "waves_effect_check() rate %d not supported", src_fmt->rate);
+		return -EINVAL;
+	}
+
+	if (src_fmt->channels != 2) {
+		comp_err(dev, "waves_effect_check() channels %d not supported", src_fmt->channels);
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "waves_effect_check() done");
+	return 0;
+}
+
+/* initializes MaxxEffect based on stream parameters */
+static int waves_effect_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+	struct comp_data *component = comp_get_drvdata(dev);
+
+	const struct audio_stream *src_fmt = &component->ca_source->stream;
+
+	MaxxStatus_t status;
+	MaxxBuffer_Format_t sample_format;
+	MaxxBuffer_Layout_t buffer_format;
+	int32_t sample_bytes;
+	MaxxStreamFormat_t *i_formats[NUM_IO_STREAMS] = { &waves_codec->i_format };
+	MaxxStreamFormat_t *o_formats[NUM_IO_STREAMS] = { &waves_codec->o_format };
+
+	comp_dbg(dev, "waves_effect_init() start");
+
+	sample_format = format_convert_sof_to_me(src_fmt->frame_fmt);
+	if (sample_format < 0) {
+		comp_err(dev, "waves_effect_init() sof sample format %d not supported",
+			 src_fmt->frame_fmt);
+		return -EINVAL;
+	}
+
+	buffer_format = layout_convert_sof_to_me(component->ca_source->buffer_fmt);
+	if (buffer_format < 0) {
+		comp_err(dev, "waves_effect_init() sof buffer format %d not supported",
+			 component->ca_source->buffer_fmt);
+		return -EINVAL;
+	}
+
+	sample_bytes = sample_format_convert_to_bytes(sample_format);
+	if (sample_bytes < 0) {
+		comp_err(dev, "waves_effect_init() sample_format %d not supported",
+			 sample_format);
+		return -EINVAL;
+	}
+
+	waves_codec->request_max_bytes = 0;
+	waves_codec->response_max_bytes = 0;
+	waves_codec->response = 0;
+	waves_codec->i_buffer = 0;
+	waves_codec->o_buffer = 0;
+
+	waves_codec->i_format.sampleRate = src_fmt->rate;
+	waves_codec->i_format.numChannels = src_fmt->channels;
+	waves_codec->i_format.samplesFormat = sample_format;
+	waves_codec->i_format.samplesLayout = buffer_format;
+
+	waves_codec->o_format = waves_codec->i_format;
+
+	waves_codec->sample_size_in_bytes = sample_bytes;
+	waves_codec->buffer_samples = src_fmt->rate * 2 / 1000; /* 2 ms io buffers */
+	waves_codec->buffer_bytes = waves_codec->buffer_samples * src_fmt->channels *
+		waves_codec->sample_size_in_bytes;
+
+	comp_info(dev, "waves_effect_init() rate %d, channels %d", waves_codec->i_format.sampleRate,
+		  waves_codec->i_format.numChannels);
+
+	comp_info(dev, "waves_effect_init() format %d, layout %d, frame %d",
```

Comment:
why a second print?

Reply:
there is a limitation of traces that allows only 4 values to be printed. I need 5. 
META_COUNT_VARAGS_BEFORE_COMPILE assertion fails if try to print 5 in one trace message

Reply:
@stolx Might be worth mentioning this in a comment.

Reply:
comment added, thanks